### PR TITLE
contrib/go-chi: customize span resource name for go-chi

### PR DESCRIPTION
**Pull request by opsengine**
### What does this PR do?
Let the caller define how the resource name should be constructed, given the `http.Request` object. If `WithResourceName` is not used, the old behavior is preserved.

### Motivation
Improve span resource names.

### Describe how to test/QA your changes
Unit tests

### Reviewer's Checklist

- [x] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

**Review by katiehockman — APPROVED** [review]
LGTM aside from the naming convention. Thanks!

File: contrib/go-chi/chi/chi_test.go
[the remark this answers is not in the document]
Comment (katiehockman) [reply contrib/go-chi/chi/chi_test.go:1]: I also unfortunately don't see a test for checking the `ext.HTTPRoute`, or the default resource name either 😢
Would it be possible for you to write those up? If not, let us know and I can help you with that.

**Comment by taowata** [comment]
Hi @opsengine,Thank you for your valuable contribution.
I noticed that your PR is targeting versions of go-chi before v5. I'm interested in seeing this enhancement extended to v5 as well. Do you plan to adapt this for chi.v5?

It's important to note that v5 already has a WithModifyResourceName function, so we'd need to ensure compatibility with this existing feature.

@katiehockman, I would appreciate your thoughts on this matter too.

**Comment by opsengine** [comment]
> It's important to note that v5 already has a WithModifyResourceName function, so we'd need to ensure compatibility with this existing feature.

I looked at `WithModifyResourceName` and it has a more limited interface because it won't give access to the `http.Request object`.

> Do you plan to adapt this for chi.v5?

Not for the moment.

**Comment by taowata** [comment]
Thanks for clarifying the limitations of `WithModifyResourceName` and your current plans regarding chi.v5.

Given the widespread usage of chi.v5, I'm considering extending this enhancement to it. If you're open to it, I'd like to create a new PR to make this possible.

**Comment by opsengine** [comment]
Yes, please go ahead :)

**Comment by opsengine** [comment]
@katiehockman good to merge?

**Review by zarirhamza — APPROVED** [review]
LGTM! Thank you @opsengine for your contribution and it sounds like a great idea @taowata to extend similar functionality to v5 as well as you mentioned [here](https://github.com/DataDog/dd-trace-go/pull/1950#issuecomment-1561380039)

**Review by ahmed-mez — APPROVED** [review]

**Pull request merged**